Reader: Add URL preview cards to posts

obenland opened the pull request:
See https://github.com/Automattic/wordpress-activitypub/issues/2157

With the upcoming addition of Fediverse RSS feeds and their short-form content character, this PR adds link preview cards to make consumption of short posts with plain URLs a richer experience.

<img width="1510" height="992" alt="image" src="https://github.com/user-attachments/assets/b20ff79b-fd45-49b6-abbd-d5c8b7745fbb" />


## Proposed Changes

* Add ContentProcessor component that scans post content for URLs.
* Add LinkPreview component that fetches and displays OpenGraph metadata.
* Integrate URL preview cards into Reader full post view.
* Support for square thumbnail images with cover fit.

## Why are these changes being made?

Modern content platforms provide rich link previews to enhance user experience and provide context about shared links. This feature brings the WordPress.com Reader in line with social media and messaging apps by automatically detecting URLs in posts and displaying preview cards with titles, descriptions, images, and metadata. This improves content discovery and engagement while maintaining the Reader's clean, focused design.

## Testing Instructions

* Navigate to a Reader post that contains URLs (either plain text URLs or links)
* Verify that preview cards appear below the post content
* Check that cards display title, description, site name, favicon, and publish date when available
* Test with posts that have images vs. no images to see different border styling
* Verify square thumbnail images display correctly with cover fit
* Check responsive behavior on different screen sizes

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

obenland committed:
Reader: Add URL preview cards to posts

Adds link preview functionality to the Reader full post view:
- Detects URLs in post content (both plain text and href attributes)
- Displays rich preview cards with OpenGraph data
- Shows title, description, site name, favicon, and publish date
- Square thumbnail images with cover fit for consistent layout
- TypeScript implementation with proper type safety
- Responsive design with rounded corners and hover states
- Conditional styling based on image availability

pfefferle commented:
I know this is still WIP, but I thought I give some early feedback.

The card looks weird if there is no image:

<img width="799" height="289" alt="Screenshot 2025-09-16 at 08 56 32" src="https://github.com/user-attachments/assets/7672e4c6-3853-43a5-a699-83a499a35c81" />

We should also only add one link preview (like bluesky, mastodon and threads), otherwise this can get very noisy if someone posts huge lists of links.

<img width="355" height="655" alt="Screenshot 2025-09-16 at 09 01 25" src="https://github.com/user-attachments/assets/8973e528-f8a7-473d-95cb-932a5510e0c1" />

We should also decide if we want to use the first or the last link and prioritize links over @-replies. I think it is fine to load profiles, but I would only use them as fallback.

<img width="858" height="448" alt="Screenshot 2025-09-16 at 08 56 49" src="https://github.com/user-attachments/assets/d7abdbc7-c1aa-4f25-8ff2-3d64ac2a549b" />

obenland commented:
@pfefferle This is great feedback, thank you!

obenland committed:
Reader: Enhance URL detection to skip mentions/hashtags and limit to first URL

- Export detectUrls function for testing
- Prioritize plain text URLs over HTML link URLs
- Skip URLs with @ or # link text (mentions/hashtags)
- Return single URL string instead of array
- Add comprehensive test coverage (19 tests)

obenland committed:
Reader: Fix HTML entity decoding in OpenGraph descriptions

- Import and use existing decodeEntities utility from lib/formatting
- Decode HTML entities in title, description, imageAlt, and siteName
- Handle entities like &#8217; ('), &amp; (&), &#8220; ("), etc.
- Add comprehensive tests for entity decoding functionality
- Remove duplicate custom decoding function

obenland committed:
Reader: Skip URL parsing for content with media elements and prioritize URL link text

- Skip URL detection when content contains img, video, or audio elements
- Prioritize HTML links where link text is a URL over regular links
- Add comprehensive tests for media detection and link prioritization
- Improve URL parsing logic with fallback mechanism

obenland committed:
Reader: Improve URL validation and link preview performance

- Add hostname validation to reject malformed URLs like https://www
- Remove problematic CORS proxy and timeout logic
- Only display previews with meaningful metadata (title + description/image/siteName)
- Hide loading states and errors - only show when preview is useful
- Fix TypeScript errors with proper event target casting
- Strip HTML tags when checking if link text looks like URL

obenland committed:
Reader: Fix tests for link preview component

- Add og:description to tests that now require meaningful metadata
- Update image URL tests to expect resolved absolute URLs
- Add iframe test to verify URL detection is skipped for iframe content

obenland committed:
Reader: Fix slow-running tests with proper fetch mocking

- Improve fetch mock setup to prevent real network calls
- Add test for failure case where component renders nothing
- Update favicon error handling test to check for display:none
- All tests now run quickly without hanging

obenland committed:
Merge branch 'trunk' into add/reader-url-preview

obenland committed:
Reader: Switch link preview to use WordPress REST API endpoint

- Replace client-side HTML parsing with server-side WordPress API call
- Use wpcom.req.get() with wpcom/v2 namespace for proper API integration
- Remove unnecessary HTML parsing, entity decoding, and URL resolution logic
- Simplify component with direct API response usage and promise-based flow
- Improve performance by eliminating CORS proxy and client-side processing

Co-Authored-By: Contributor <noreply@github.com>

obenland committed:
Reader: Update link preview tests for WordPress REST API

- Replace fetch mocks with wpcom.req.get mocks to match new API implementation
- Remove HTML parsing test scenarios since processing now happens server-side
- Simplify test data structure to match API response format
- Add test coverage for WordPress REST API parameter validation
- All tests passing with new server-side architecture

obenland committed:
Reader: Refine URL detection and add loading state

- Comment out fallback for URLs with regular anchor text in content processor
- Only process plain URLs and URLs where link text matches the URL
- Add loading spinner component to show preview fetch state
- Update tests to match new URL detection behavior
- All 55 tests passing with refined functionality

obenland committed:
Reader: Use proper query parameters for wpcom API call

- Replace manual URL encoding with proper wpcom.req.get parameter format
- Pass URL as query parameter object instead of encoding in path
- Cleaner and more maintainable API call pattern

obenland committed:
Reader: Update link preview tests for API endpoint change

- Update test expectations to match new /read/link-preview endpoint path
- Fix API call parameters to use two-argument format with query object
- All 55 tests passing with updated endpoint structure

obenland committed:
Update link preview API path

Changed the API path from '/link-preview' to '/read/link-preview' in LinkPreview component to match the updated endpoint.

pfefferle commented:
I like that it’s limited to showing only URL links! That keeps the view focused on microblogging, and we can always monitor it to consider opening up to other link types later.

It works well overall, though the restriction `array( 'limit_response_size' => 200 * KB_IN_BYTES )` causes some sites like YouTube not to load properly.

jeherve reviewed (commented):
Would it be an option to replace the link by the preview, instead of adding the preview at the end of the post? If one posts a link and then some content, it's a bit confusing to see the link, the content, and then a preview on its own, outside of its context, at the end of the post.

Replacing the link with the post preview would allow displaying multiple previews in a single post when there are multiple links in a single post.

jeherve commented:
> Are there any examples where you thought it wasn't a good fit?

I don't have great examples right now, so it may turn up to be more of an edge case we don't need to worry about.

obenland committed:
Merge branch 'trunk' into add/reader-url-preview

obenland committed:
Merge branch 'trunk' into add/reader-url-preview
